extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -386,16 +385,58 @@ public void run()
             records = consumer.poll(POLL_TIMEOUT);
           }
           catch (OffsetOutOfRangeException e) {
-            log.warn("OffsetOutOfRangeException with message [%s], retrying in %dms", e.getMessage(), POLL_RETRY_MS);
-            pollRetryLock.lockInterruptibly();
-            try {
-              long nanos = TimeUnit.MILLISECONDS.toNanos(POLL_RETRY_MS);
-              while (nanos > 0L && !pauseRequested && !stopRequested) {
-                nanos = isAwaitingRetry.awaitNanos(nanos);
+            // Reset consumer offset if resetOffsetAutomatically is set to true
+            // and the current message offset in the kafka partition is more than the
+            // next message offset that we are trying to fetch
+            if (tuningConfig.isResetOffsetAutomatically()) {
+              for (TopicPartition topicPartition : consumer.assignment()) {
+                final long currentOffset = consumer.position(topicPartition);
+                log.trace("Current consumer position is [%d]", currentOffset);
+                log.trace("Next offset required is [%d]", nextOffsets.get(topicPartition.partition()));
+                // seek to the beginning to get the least available offset
+                consumer.seekToBeginning(topicPartition);
+                final long leastAvailableOffset = consumer.position(topicPartition);
+                log.trace("Least consumer offset is [%d]", leastAvailableOffset);
+                // reset the seek
+                consumer.seek(topicPartition, currentOffset);
+
+                if (leastAvailableOffset > currentOffset) {
+                  if (ioConfig.isUseEarliestOffset()) {
+                    log.makeAlert("Got OffsetOutOfRangeException [%s], resetting to the earliest offset", e.getMessage());
+                    consumer.seekToBeginning(topicPartition);
+                    nextOffsets.put(topicPartition.partition(), consumer.position(topicPartition));
+                    log.warn("Consumer is now at offset [%d]", consumer.position(topicPartition));
+                  } else {
+                    log.makeAlert("Got OffsetOutOfRangeException [%s], resetting to the latest offset", e.getMessage());
+                    consumer.seekToEnd(topicPartition);
+                    nextOffsets.put(topicPartition.partition(), consumer.position(topicPartition));
```

Comment:
These lines are repeated and can be pulled out of the `if (ioConfig.isUseEarliestOffset())`